text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1435 @@
+to resources (such as textures) and these types, like wasm's `externref`, cannot
+be put in memory.
+
+[^9]: When Rust is compiled to wasm, we can think of the memory of the Rust program
+as being backed by something like a `[u8]`, `externref`s exist outside of that `[u8]`
+and there is no way to put an `externref` into this memory, so it is impossible to have
+a reference or pointer to a `externref`. `wasm-bindgen` currently supports `externref`
+by creating a array of the items which would be referenced by an `externref` on the
+host side and passes indices into this array across the wasm-host boundary in lieu
+of `externref`s. It isn't possible to support opaque types from some GPU targets using
+this technique.
+
+`externref` are similar to `Pointee` in that the type's size is not known, but unlike
+`Pointee` cannot be used behind a pointer. This RFC's proposed hierarchy of traits could
+support this by adding another supertrait, `Value`:
+
+```
+    ┌────────────────┐                  ┌─────────────────────────────┐
+    │ const Sized    │ ───────────────→ │ Sized                       │
+    │ {type, target} │     implies      │ {type, target, runtime env} │
+    └────────────────┘                  └─────────────────────────────┘
+            │                                          │
+         implies                                    implies
+            │                                          │
+            ↓                                          ↓
+┌───────────────────────┐             ┌────────────────────────────────────┐
+│ const ValueSized      │ ──────────→ │ ValueSized                         │
+│ {type, target, value} │   implies   │ {type, target, runtime env, value} │
+└───────────────────────┘             └────────────────────────────────────┘
+                                                       │
+                                                    implies
+                                                       │
+                               ┌───────────────────────┘
+                               ↓
+                     ┌──────────────────┐
+                     │ Pointee          │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+                               │
+                            implies
+                               │
+                               ↓
+                     ┌──────────────────┐
+                     │ Value            │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+```
+
+`Pointee` is still defined as being implemented for any type that can be used
+behind a pointer and may not be sized at all, this would be implemented for
+effectively every type except wasm's `externref` (or similar opaque types from
+some GPU targets). `Value` is defined as being implemented for any type that can
+be used as a value, which is all types, and also may not be sized at all.
```

Comment:
One possible complication: Type which are `Value` (but not `Pointee`, referred to as `!Pointee` types from now on) would be _**very**_ restrictive in how they can be used (without additional, IMO somewhat significant, changes to the language).

If a type is neither `Copy` nor allows references to it to be taken, then it can be basically used in a single function and passed into a single function, but not much more (unless it's returned again by the called function, which might work as long as you only need to return a single value).

To make any real use of`!Pointee` types, I think the language would need to support at least one (or more) of the following:

1) Allow returning multiple `!Pointee` values from a function, potentially by supporting `!Pointee` types in structs / tuples (which would require passing each field of such aggregates individually at the ABI level).

2) Allow such `!Pointee` types to be copied, either by somehow allowing them to implement `Copy` (despite the conflicting super traits), or allowing them to be copied even if they do not implement the `Copy` trait.

3) Allow **immutable** references to `!Pointee` values to exist, but make these themselves `!Pointee`. At the ABI level such references would be represented by the value itself. `UnsafeCell` would not be allowed to contain `!Pointee`.

(3) would be the most flexible solution here, though also the largest addition to the language semantics, I think.

Anyway, the details here are out of scope of this RFC, I think, but to me this seems like a significant enough complication that it might be worthwhile to mention here, or to weaken the wording ("proposed hierarchy of traits **could support** this", "it would be **possible to support**") a bit, to not imply that this would solve all the `externref` problems. (Maybe adding something along the lines of "though making proper use of `!Pointee` types could require additional additions to the language" somewhere might be sufficient for this RFC).

Reply:
I've been thinking more about this (see also https://hackmd.io/@TimNN/Hk1LO6iMJe) and I think one thing that doesn't work well at all is the modeling of the "can be stored in a local variable" property.

We currently have four groups of types that could potentially be stored in local variables:

* Sized values
* Dynamically-sized values
* Unsized values (i.e. `extern` types)
* Abstract values, as I have taken to calling them (e.g. `externref`).

`Sized` values are fully supported.

Dynamically-sized values are support via `unsized_locals` and `unsized_fn_params`. They use a variable amount of storage at runtime. They (currently) cannot be used as a function return value.

Unsized values conceptually cannot be stored in local variables. (This RFC currently proposes that  _"it could instead be permitted to write functions which use extern types as values, and instead rely on it being impossible to get a extern type that is not behind a pointer or a reference"_, but that doesn't seem like a great solution to me).

Abstract values should be usable in local variables, function arguments, and as function return values. Contrary to dynamically-sized values, they do not use a variable amount of storage at runtime (but have other restrictions).

---

I don't have a good solution right now, especially since I'm not sure about what the long-term plan for dynamically-sized values is (should they be usable as function return values?).

However I'm currently feeling that the "can be stored in a local variable" property might not fit very well into the hierarchy proposed here.

If we assume that dynamically-sized values support all operations that are supported for abstract values (including being usable as a return value), then it might be reasonable to define `ValueSized: Pointee + AbstractValue`, with no direct relationship between `Pointee` and `AbstractValue`

(Though if `Pointee` and `AbstractValue` exist independently, then the question becomes "what does `: ?ValueSized` mean?". Depending on the outcome of other discussions, this may mean bounds such as`: ?ValueSized + AbstractValue`).

--- 

One way I've been trying to think about this is, what would be appropriate bounds for `V, R and Self` in the following traits?

```rs
trait FromVal<V> { fn from_val(_: V) -> Self }

trait FromRef<R> { fn from_ref(_: &R) -> Self }
```